### PR TITLE
Popular Repos Page

### PR DESCRIPTION
**Pull request by aidan-epperly**
This PR adds the popular repos page which is currently documented in #403

### CODE REVIEW

**Comment by aidan-epperly** [comment]
image

image

**Comment by LRWeber** [comment]
I think it would be easier to read if we change the Creation History tips to put one repo name per line for shared dates, similar to what the Pull Requests and Issues graphs do now.

Is there a particular reason for using a top 8? There's definitely an apparent top 8 currently in the bubbles/clusters, but that's likely to chinge over time. It just seems like an unusual number.

Regardless of the number we pick, we should probably include some indication in the titles. (e.g. Instead of "Activity Across Most Popular Repos by Stars" say something like "Activity Across Top 8 Repos by Stars")

**Comment by aidan-epperly** [comment]
>I think it would be easier to read if we change the Creation History tips to put one repo name per line for shared dates, similar to what the Pull Requests and Issues graphs do now.

Alright, I can do that.

>Is there a particular reason for using a top 8? There's definitely an apparent top 8 currently in the bubbles/clusters, but that's likely to chinge over time. It just seems like an unusual number.

Not particularly. I just averaged 5 and 10 because that was the range that I had talked about. Right now it is a top 'n' list, but I could change it to be representative of something like percentage of total stars?

>Regardless of the number we pick, we should probably include some indication in the titles. (e.g. Instead of "Activity Across Most Popular Repos by Stars" say something like "Activity Across Top 8 Repos by Stars")

Ok, I can do that pretty easily.

**Comment by aidan-epperly** [comment]
How's this?
image

**Comment by LRWeber** [comment]
That works!
Can we add some kind of Top 10 note to the Creation History as well? e.g. "Repo Creation History, Top 10 Repos by Stars"

**Comment by aidan-epperly** [comment]
So, that is all done. I also added a modified license graph only showing the licenses for the top repos by stars. Anything else you think I should add before taking this off of draft? I thought about adding pie charts to show the relative total size, but I don't think that that is really necessary.

**Comment by LRWeber** [comment]
The spacing between the first and second graph feels a bit excessive. Other than that, it looks great!

File: js/explore/largeRepos/sunburst_licenses.js [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+            }
+            if (objTree.hasOwnProperty(licenseName)) {
+                objTree[licenseName][repo] = repo;
+            } else {
+                objTree[licenseName] = {};
+                objTree[licenseName][repo] = repo;
+            }
+        }
+
+        // Converts object tree into d3.hierarchy complient form
```
Comment (LRWeber) [inline js/explore/largeRepos/sunburst_licenses.js:308]:
```suggestion
        // Converts object tree into d3.hierarchy compliant form
```
Same typo also in [`js/explore/sunburst_licenses.js`, L305](https://github.com/LLNL/llnl.github.io/blob/c72139812740b49d7073afc88ff9d373daaf9cb2/js/explore/sunburst_licenses.js#L305)

File: js/explore/largeRepos/sunburst_licenses.js [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+                        });
+                path.selectAll('title').remove();
+                path.append('title').text(d => d.data.name);
+            });
+    
+        // Creates option slider
+        /*chart.append('g')
+            .attr('transform', `translate(${margin.left},${margin.top / 2})`)
+            .attr('id', 'licenseSlider')
+            .call(slider);*/
```
Comment (LRWeber) [inline js/explore/largeRepos/sunburst_licenses.js:276]: I think the option slider would still be useful for the top 10. We are still showing the "Top 10 Repos by Stars", but it still makes sense to change their relative proportions based on other factors.
```suggestion
        // Creates option slider
        chart.append('g')
            .attr('transform', `translate(${margin.left},${margin.top / 2})`)
            .attr('id', 'licenseSlider')
            .call(slider);
```

Comment (aidan-epperly) [reply js/explore/largeRepos/sunburst_licenses.js:276]: Alright, sounds good. I didn't know how it would look to someone else, so good to know.

**Comment by tzanio** [comment]
This looks really nice thanks for adding it!

Is it possible to have a version where instead of number of commits, you plot the numbers of lines changed? Commits can be of different sizes so this will be more representative of the amount of "work".

It could also be interesting to normalize the above by the current total number of lines in the repo, as a measure of how much the project has changed overall.

**Comment by aidan-epperly** [comment]
Glad that you like it!

Unfortunately, as far as I know, the GitHub API does not have that information, though I am only very familiar with the GraphQL API, and not the old REST API. I agree that it would be interesting, but I don't know if it would be feasible for me to implement. Laura Weber is far more familiar with the API and what it can do though, so she is probably the better person to answer this.

**Comment by IanLee1521** [comment]
Yeah, to get that granular, I think you would have to clone every repo down and then go through the commit history there... It's doable, but it's a bit of work.

**Comment by LRWeber** [comment]
The activity we've been collecting has only ever come from the REST API
And it turns out... we might actually have that info readily available.
https://developer.github.com/v3/repos/statistics/#get-the-weekly-commit-activity

**Comment by aidan-epperly** [comment]
Yeah, that would be great! I don't know how to change the REST API query, but I could definitely change the graph to show that data if we had it.

**Comment by IanLee1521** [comment]
Right, that works for commits, but do those commits give us the number of lines of code changed., which is what @tzanio was asking about.

**Comment by aidan-epperly** [comment]
I believe [this](https://developer.github.com/v3/repos/statistics/#response-2) seems to suggest that.

**Comment by aidan-epperly** [comment]
## Response
Returns a weekly aggregate of the number of additions and deletions pushed to a repository.

```
Status: 200 OK
[
  [
    1302998400,
    1124,
    -435
  ]
]
```

**Comment by IanLee1521** [comment]
ah nice. So that would get us new data going forward, though might not get us the history for the last year.

**Comment by LRWeber** [comment]
From https://developer.github.com/v3/repos/statistics/#get-the-weekly-commit-activity

> Get the weekly commit activity
> `GET /repos/:owner/:repo/stats/code_frequency`
> Response
> Returns a weekly aggregate of the number of additions and deletions pushed to a repository.
> 
> ```
> Status: 200 OK
> [
>   [
>     1302998400,
>     1124,
>     -435
>   ]
> ]
> ```

**Comment by LRWeber** [comment]
> ah nice. So that would get us new data going forward, though might not get us the history for the last year.

I believe that is what it is, but I'd have to run the query to be sure.

**Comment by aidan-epperly** [comment]
There is also the deeper question of whether or not additions/deletions is a more accurate than commits given that, for instance, the updating of JSON files in this project is responsible for over a million changes in total.

**Comment by IanLee1521** [comment]
heh, yeah, I'm probably inclined to say that commits is the right metric, but that's just me

**Comment by aidan-epperly** [comment]
> None of these are perfect metrics though.

True, though if we had all the data, I could actually add a slider that changes how highly we weight a given metric.

**Comment by LRWeber** [comment]
I think the right answer for "most accurate metric" varies by project, depending on their particular workflow.

We could collect data for both. It looks like the same kind of year-long per-week data to me!
https://api.github.com/repos/LLNL/llnl.github.io/stats/code_frequency

**Comment by IanLee1521** [comment]
I wonder if we could try to calculate a "churn" metric for this...

Side note... I wonder if this discussion should move to a new issue, rather than being on a merged PR?

**Comment by aidan-epperly** [comment]
> Side note... I wonder if this discussion should move to a new issue, rather than being on a merged PR?

Sure, I can make one now.

**Comment by tzanio** [comment]
Thanks so much for the quick response everyone!

**Pull request merged**